magenta-lib/src/main/scala/magenta/tasks/UpdateCloudFormationTask.scala

```diff
@@ -242,6 +267,35 @@ trait CloudFormation extends AWS {
     client.describeStackEvents(
       new DescribeStackEventsRequest().withStackName(name)
     )
+
+  def findStackByTags(tags: Map[String, String], reporter: DeployReporter)(implicit keyRing:KeyRing) = {
+
+    def tagsFilter(stack: AmazonStack): Boolean =
+      tags.forall { case (key, value) => stack.getTags.exists(t => t.getKey == key && t.getValue == value) }
+
+    @tailrec
+    def recur(nextToken: String = null, existingStacks: List[AmazonStack] = Nil): List[AmazonStack] = {
+      val request = new DescribeStacksRequest().withNextToken(nextToken)
+      val response = client.describeStacks(request)
+
+      val stacks = response.getStacks.foldLeft(existingStacks) {
+        case (agg, stack) if tagsFilter(stack) => stack :: agg
+        case (agg, _) => agg
+      }
+
+      if (response.getNextToken == null)
+        stacks
+      else
+        recur(response.getNextToken, stacks)
+    }
+
+    recur() match {
+      case cfnStack :: Nil => Some(cfnStack)
+      case Nil => reporter.fail(s"No matching cloudformation stack match for $tags.")
+      case _ =>
+        reporter.fail(s"More than one cloudformation stack match for $tags. Failing fast since this may be non-deterministic.")
+    }
```

Comment:
Part of me wonders if this logic should be outside of this function to make it less dependent on side-effects. But I'm aware that much of riff-raff is like this so I'm not too bothered.